topic-operator/src/main/java/io/strimzi/operator/topic/Session.java

```diff
@@ -171,96 +212,116 @@ public void start(Promise<Void> start) {
-            zkResult -> {
-                if (zkResult.failed()) {
-                    start.fail(zkResult.cause());
-                    return;
-                }
-                this.zk = zkResult.result();
-                LOGGER.debug("Using ZooKeeper {}", zk);
-
-                String topicsPath = config.get(Config.TOPICS_PATH);
-                TopicStore topicStore;
-                if (config.get(Config.USE_ZOOKEEPER_TOPIC_STORE)) {
-                    topicStore = new ZkTopicStore(zk, topicsPath);
-                } else {
-                    boolean exists = zk.getPathExists(topicsPath);
-                    CompletionStage<KafkaStreamsTopicStoreService> cs;
-                    if (exists) {
-                        cs = Zk2KafkaStreams.upgrade(zk, config, adminClientProperties(), false);
-                    } else {
-                        KafkaStreamsTopicStoreService ksc = new KafkaStreamsTopicStoreService();
-                        cs = ksc.start(config, adminClientProperties()).thenCompose(s -> CompletableFuture.completedFuture(ksc));
-                    }
-                    topicStore = ConcurrentUtil.result(
-                            cs.handle((s, t) -> {
-                                if (t != null) {
-                                    LOGGER.error("Failed to create topic store.", t);
-                                    start.fail(t);
-                                    return null; // [1]
-                                } else {
-                                    service = s;
-                                    return s.store;
-                                }
-                            }));
-                    if (topicStore == null) {
-                        return; // [1]
-                    }
-                }
+        startHealthServer(topicOperatorState)
+                .onFailure(cause -> LOGGER.error("Failed to start health server", cause))
+                .onSuccess(httpServer -> healthServer = httpServer)
+                .compose(ignored -> zkClientCreator.apply(vertx, config))
+                .onSuccess(zookeeper -> zk = zookeeper)
+                .compose(zk -> executor.executeBlocking(createTopicStoreAsync(zk, config)))
+                .onSuccess(topicStore -> LOGGER.debug("Using TopicStore {}", topicStore))
+                .compose(topicStore -> createTopicOperatorAndZkWatchers(labels, namespace, topicStore))
+                .compose(this::createK8sWatcher)
+                .onSuccess(this::createPeriodicReconcileTrigger)
+                .onSuccess(ignored -> {
+                    start.complete();
+                    topicOperatorState.setReady(true);
+                    LOGGER.info("Started");
+                })
+                .onFailure(cause -> {
+                    topicOperatorState.setAlive(false);
+                    start.fail(cause);
+                    LOGGER.error("Topic operator start up failed, cause was", cause);
+                });
+    }
```

Comment:
I had to refactor this chain due to my misunderstanding of how Vert.x futures work - basically, if a future fails, the failed future is passed along. 

Previously when I had multiple `onFailure` clauses, they were triggering when an upstream future failed, causing misleading log messages (e.g., the ZK client failing to create would trigger its `onFailure` but then also trigger the `onFailure` clause that followed the topic store creation).

So now I rely on the fact that the argument to a single arg `compose` call is only ever called when the previous future is successful, so will never trigger for a failed previous future.

So I've a) refactored some onSuccess/onFailure handlers out to join their parent future in their own method so we still log on success/failure, but having it in a separate method keeps this chain cleaner. 
and b) as a failed future will be passed all the way down the chain, all of the code around failing the start up promise, and setting the topicOperatorState liveness, can be handled in one onFailure at the very end.

Reply:
There vertx Javadocs are pretty good, but I've always thought it's pretty easy to miss the significance of the word 'this' (meaning the method receiver) in those docs. Note also there's a two parameter version of `compose` which takes one handler for success and another for failure, which _can_ be clearer, but not if you're passing lambdas where the fact that there are two arguments is easily lost. If both args are method references then it's fine.

Reply:
Yeah, I was looking at the two param compose, but as you said, it made things harder to read with lambdas.